Added the ability to define tags for elasticache

memo42 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Added the ability to add tags. Fixes #24214
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
elasticache

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (devel 3142e60693) last updated 2017/08/09 11:02:06 (GMT +200)
  config file = /home/mo/playbooks/production.cfg
  configured module search path = [u'/home/mo/playbooks/library']
  ansible python module location = /home/mo/ansible/lib/ansible
  executable location = /home/mo/ansible/bin/ansible
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

memo42 committed:
added tags to aws elasticache. Fixes #24214

memo42 committed:
added example for how to add tags

memo42 committed:
removed whitespace markers

memo42 committed:
removed more whitespace markers

memo42 committed:
fixed version added

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -128,6 +135,25 @@
+    name: "test-please-delete"
+    state: present
+    engine: redis
+    tags: [
```
This should use standard ansible tag form

```
tags:
  test: testa
  testb: testc
  testd: testf
```

You'll want to use `ansible_dict_to_boto3_tag_list` for the conversion

memo42 replied:
OK, thanks

willthames reviewed (changes requested):
There is a lot of work involved in managing tags. Have a look at e.g. ec2_group.

In particular:
* Adding tags on new resources
* Updating tags on existing resources
* Tracking `changed` if a tag gets added or amended
* Using `purge_tags` to remove tags only if set

Have a look at `compare_aws_tags` too to manage tag amendment.

I also strongly recommend adding test cases around this too. Again, ec2_group is a good model.

memo42 committed:
made elasticache work with tags given as ansible dict

memo42 committed:
fixed bug where state was set to unmodifiable as soon as a cache with the same name existed

memo42 committed:
implemented updating tags

memo42 committed:
fixed pep8 issues

memo42 committed:
now also setting changed when tags are updated

memo42 committed:
added purge_tags flag

memo42 committed:
now working better when no tags are given

memo42 committed:
fixed bug where changed was set without changes when no num_nodes or cache_engine_version was set

memo42 commented:
So your requests should be met now. I tried getting started with tests, but ran into bigger problems ( Issue #28351 ), so that might take a litte bit longer.

willthames commented:
Sorry, I've only just noticed this again (was looking for some other elasticache features). I'll endeavour to test this out shortly

memo42 commented:
Nice, thanks. I did not have time to properly get into your test-setup, yet. I did get the local tests running, but there was no security group template for elasticache, my first, hacky attempt of building one did not work and I was not able to try again since.

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -182,6 +210,37 @@ def ensure_present(self):
+            self.module.fail_json(msg="region must be specified")
+        instance_counts = {}
+        response = client.describe_db_instances()
+        user_id = response['DBInstances'][0]['DBInstanceArn'].split(':')[4]
```
Argh, it's so frustrating that the Elasticache API doesn't return ARNs!

This is probably the best way to get user ID though:
https://stackoverflow.com/a/10198870/3538079

There is always a chance that a user allowed to use Elasticache isn't allowed to use `rds.describe_db_instances`, so it's best to rely on as few other APIs as possible (I think everyone has permission to run `iam.get_caller_identity`)

memo42 replied:
cool, thanks

willthames commented:
Please can you commit your tests too (even if they're failing, I can always look at improving them)

willthames commented:
Also, if you do have problems setting up tests that my blog post doesn't help with, let me know so that I can update my blog post! (Based on #28351 I should document that awscli is required!)

memo42 committed:
started writing tests for elasticache module

memo42 committed:
updated version_added after 2.4 release

memo42 committed:
self.tags is now always an ansible dict

memo42 committed:
improved arn calculation using more commonly accessable functions

memo42 committed:
now only removing tags if purge_tags is set

memo42 committed:
fixed pep8 issues in elasticache module

memo42 commented:
Thanks for your help.
I fixed the getARN method the way you suggested.
I merged in the current devel branch, which ansibot did not like so I rebased and force pushed. That is why this PR now has some weird tags. Sorry about that.
I also pushed my tests. I did not get very far, simply copied the tests from some other module, removed everything that was irrelevant and adjusted a few details.
The problem was that the playbook that created the iam policy did not have a template for elasticache permissions. I tried adding one that allows all for a first test, but it did not work.
I probably just need to spend some time with this to figure everything out.
Not sure when I will get to that, though.

yuwzho commented:
-label azure

dharmabumstead commented:
Closing this one. Feel free to rebase and reopen.

willthames commented on hacking/aws_config/testing_policies/elasticache-policy.json:
```diff
@@ -0,0 +1,15 @@
+{# Note that not all EC2 API Actions allow a specific resource #}
```
Please don't add a new policy - reuse storage or compute for elasticache (we can only attach 10 policies to a group, so we need to limit the number we have)

willthames commented on hacking/aws_config/testing_policies/elasticache-policy.json:
```diff
@@ -0,0 +1,15 @@
+{# Note that not all EC2 API Actions allow a specific resource #}
+{# See http://docs.aws.amazon.com/AWSEC2/latest/APIReference/ec2-api-permissions.html#ec2-api-unsupported-resource-permissions #}
```
this comment is unrelated to elasticache

willthames commented on hacking/aws_config/testing_policies/elasticache-policy.json:
```diff
@@ -0,0 +1,15 @@
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Sid": "AllowUnspecifiedEC2Resource",
```
Please update the SID for Elasticache

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -95,6 +95,21 @@
+      - Set instance tags
+    required: false
+    default: none
+    version_added: "2.5"
```
needs to be 2.6 now

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -182,6 +210,35 @@ def ensure_present(self):
+        return arn
+
+    def compare_and_update_tags(self):
+        if self.tags:
```
This check shouldn't be required if `self.tags` defaults to the empty dict - the else clause would be taken care of in the same way using compare_aws_tags.

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -386,7 +447,7 @@ def _requires_modification(self):
         }
         for key, value in modifiable_data.items():
-            if value is not None and self.data[key] != value:
+            if value and self.data[key] != value:
```
the change from `value is not None` is very dangerous if any value can be `False` instead.

willthames commented on lib/ansible/modules/cloud/amazon/elasticache.py:
```diff
@@ -420,7 +481,7 @@ def _requires_destroy_and_create(self):
             unmodifiable_data['zone'] = self.data['PreferredAvailabilityZone']
         for key, value in unmodifiable_data.items():
-            if getattr(self, key) is not None and getattr(self, key) != value:
+            if getattr(self, key) and getattr(self, key) != value:
```
Again, this can be dangerous if any value can be `False`
